### PR TITLE
[JENKINS-69955] Make websocket connection idleTimeout configurable

### PR DESCRIPTION
**Pull request by Dohbedoh**
and adjust the default

See [JENKINS-69955](https://issues.jenkins.io/browse/JENKINS-69955).

Since Jetty 10, websocket connections have a [default idle timeout of 30s](https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConstants.java#L32). As opposed to a default [idle timeout of 5 minutes](https://github.com/eclipse/jetty.project/blob/jetty-9.4.48.v20220622/jetty-websocket/websocket-api/src/main/java/org/eclipse/jetty/websocket/api/WebSocketPolicy.java#L81-L86) in Jetty 9. 

This does not play well with the current ping interval mechanism that is used to keep connections active. The ping interval is [30 seconds by default](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/websocket/WebSocketSession.java#L58). Just a slight drift between 2 pings is enough have an existing websocket connection that expires..

Hence the proposal here to:

* set the websocket idle timeout of the Jetty websocket container to a reasonable default of 2 x the ping interval (60 seconds)
* allow adjusting the idle timeout via a system property `jenkins.websocket.idleTimeout`

### Testing done

**Test 1**

* Start Jenkins
* Create a websocket agent
* Connect the websocket agent
* Wait for minutes / an hour to make sure that the connection is stable

**Test 2: test of system property**

* Start Jenkins with `-Djenkins.websocket.idleTimeout=15`
* Create a websocket agent
* Connect the websocket agent
* Wait for up to 30s to see that  the connection fails (because clearly the idle timeout is lower than the default ping interval)

### Proposed changelog entries

- Adjust websocket Idle Timeout to 60s seconds by default to avoid "WebSocketTimeoutException: Connection Idle Timeout" issues. Idle timeout is configurable via `jenkins.websocket.idleTimeout=<timeoutInSeconds>`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jglick @olamy 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7670"></a>

### CODE REVIEW

File: websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
Comment (olamy) [inline websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:56]: ideally we should have here a default value linked to Long.getLong("jenkins.websocket.pingInterval").
maybe something such `private static long IDLE_TIMEOUT_SECONDS = Long.getLong("jenkins.websocket.idleTimeout", Long.getLong("jenkins.websocket.pingInterval", 30L) * 2);` 
? 
but I agree maybe sounds a bit complicated at least we have now a better default value.

**Review by olamy — APPROVED** [review]

File: websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
Comment (Dohbedoh) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:56]: I agree with this. They better be correlated. Though changing the default would require to change a value in 2 locations. We already read the property in `WebSocketSession` in `core/`.

I looked into this actually but and it is not straighforward to do it without replicating a `Long.getLong("jenkins.websocket.pingInterval", 30L)` like you suggest here. The jetty implementation is abstracted from core. Core is not accessible from `websocket/`. I thought about adding a method to the `Provider` interface but it looks like the `IdleTimeout` is something specific to Jetty, and not necessarily to other implementations of websocket..

Or maybe we could have `public static` constants in the websocket/spi to share this information.

Comment (rsandell) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:56]: Ideally it should use `jenkins.util.SystemProperties`

**Review by rsandell — APPROVED** [review]

File: websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
Comment (jglick) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:56]: > it should use `jenkins.util.SystemProperties`

It cannot do so from this module.

Comment (jglick) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:56]: Linking default values seems too complicated. In the unlikely case you find a need to customize these values, it is not a burden to set two properties.

File: websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
Comment (jglick) [inline websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:60]: Caching this object is a change of behavior. Can I assume you tested repeated agent connections within a single controller session?

**Review by jglick — APPROVED** [review]
Thanks!

File: websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
Comment (olamy) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:60]: usually this sort of configuration is done via `ServletContainerInitializer` mechanism.

Comment (olamy) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:60]: just an idea about those fields (same for `WebSocketServletFactory` in the `Jetty9Provider` class. 
we could change to something such (not tested :) )
```
Interface Provider

init(ServletContext svltCtxt)

```

`Websockets`

```
@MetaInfServices(ServletContainerInitializer.class)
public class WebSockets implements ServletContainerInitializer {

    public void onStartup(Set<Class<?>> set, ServletContext servletContext) throws ServletException {
        ServiceLoader.load(Provider.class).stream().forEach(service -> service.get().init(servletContext));
    }

```

`Jetty10Provider` 

```
   public void init(ServletContext servletContext) {
        JettyWebSocketServerContainer.ensureContainer(servletContext).setIdleTimeout(Duration.ofSeconds(IDLE_TIMEOUT_SECONDS));
    }
```
similar change for `Jetty9Provider`.


with this we do not store anymore anything in a field and do not use anymore some `synchronized` method

Comment (Dohbedoh) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:60]: > Can I assume you tested repeated agent connections within a single controller session?

Yes I did. Keeping a reference here is probably not needed. What @olamy proposes seems more consistent, to have an one time initializer on startup.

Comment (Dohbedoh) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:60]: Tested it in a separate branch https://github.com/jenkinsci/jenkins/commit/9ca7cb3e1f192b0d6c526fcd95c9108d7c85c9b0. And that seems to work just fine. I haven't tested for Jetty9 though.
@jglick let me know what you think. Using a `ServletContainerInitializer` looks good. Also how should I test the Jetty9 provider ? just remove the dependency on `websocket-jetty10` from the war/pom ?

Comment (jglick) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:60]: Since this PR is a fix for an important regression, which I have taken the liberty of marking `lts-candidate` for possible backport, let us keep the diff simple. Structural refactorings for stylistic purposes can be proposed later in trunk.

Comment (Dohbedoh) [reply websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:60]: Got it. So maybe I can use an `AtomicBoolean` to handle the single initialization and not cache the `JettyWebSocketServerContainer`.

File: websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
Comment (jglick) [inline websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java:70]: This is not actually atomic—you might as well use a simple `boolean`. I doubt it matters since calling `setIdleTimeout` twice should be harmless.

**Review by jglick — COMMENTED** [review]
FWIW I found this clearer before 695cd3c5bb847c7b44da430a9e88b87ae6c555d7. What I meant by https://github.com/jenkinsci/jenkins/pull/7670#discussion_r1116931134 is that https://github.com/jenkinsci/jenkins/pull/7670#discussion_r1116580445 would make for a much bigger diff and feels riskier, whereas the original b53abb74fa7ab5e8ed546085be7f541e7370c7ee seemed pretty straightforward and probably safe given

>> Can I assume you tested repeated agent connections within a single controller session?

> Yes I did.

I do not mean to prolong the discussion about this important fix too much; really anything along these lines is probably fine, if it has been tested successfully. @olamy any concerns about incorrectness, like subtle problems with Jetty keeping around stale objects etc.?

**Review by olamy — APPROVED** [review]

**Comment by olamy** [comment]
LGTM eventually this will setIdleTimeout multiple time to the same value. but no problem at all it's dynamic in Jetty

**Review by jglick — APPROVED** [review]

**Comment by Dohbedoh** [comment]
not sure who to ping to get this merge asap ? Maybe @basil ? This is an LTS candidate so hoping this can make it to the 2.387.1 LTS.

**Comment by NotMyFault** [comment]
> not sure who to ping to get this merge asap ?

The master branch has been frozen for next week's security updates. PRs will be on hold until then. 

> This is an LTS candidate so hoping this can make it to the 2.387.1 LTS.

2.387.1 releases next week's Wednesday already. This PR can make it into 2.387.2 onwards only.

**Comment by Dohbedoh** [comment]
I see thanks!

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by olamy** [comment]
@timja looks to be some build infra issue here...

**Comment by timja** [comment]
see https://github.com/jenkinsci/jenkins/commit/52c58e2446b16a007ee895a37a60e7e5bd739407 

and
https://github.com/jenkinsci/jenkins/pull/7700#discussion_r1131365900

it's from the security fixes, (they're aware)

**Pull request merged**
